Create Interoperability & crosschain-communication pages - Closes #1516 & #1496 (#1623)
LiskHQ/lisk-docs: docs/modules/ROOT/pages/understand-blockchain/interoperability/index.adoc

= Lisk Interoperability
:toc: preamble
// URLs
:url_yt_lisk: https://www.youtube.com/c/LiskHQ
:url_blog_research: https://lisk.com/blog/archive?field_blog_category%5B166%5D=166
:url_blog_interop_intro: https://lisk.io/blog/research/introduction-blockchain-interoperability
:url_wiki_topologies_star: https://en.wikipedia.org/wiki/Network_topology#Star
:url_understand_tx: understand-blockchain/index.adoc#transactions
:url_understand_command: understand-blockchain/sdk/modules-commands.adoc#commands
// footnotes
:fn_wiki_topology: footnote:topology[See {url_wiki_topologies_star}[^] for more information about network topologies.]
:fn_tx: footnote:tx[See xref:{url_understand_tx}[Understand blockchain, Transactions.] for more information about transactions.]
:fn_command: footnote:command[See xref:{url_understand_command}[Modules and commands]. for more information about commands.]

> 

Reviewer: Adding comment here. CCM image is wrong, should use updated schema, see: https://github.com/LiskHQ/lips/blob/update-interoperability-LIPs/proposals/lip-0049.md#cross-chain-message-schema

Author: @AndreasKendziorra do you have the original image somewhere, so I can update it?

Author: I see thanks! 

I think I will try to recreate and update the diagrams in Lucidchart

Reviewer 2: 1. The name of the cc-transfer was recently changed to "transferCrossChain". See [here](https://github.com/LiskHQ/lips/blob/ce9b653f7210f0d997a599486243ec0d031ed30b/proposals/lip-0051.md?plain=1#L133)
2. There is no `sender` property in the parameter of the cc-transfer command. But you could add `senderPublicKey: <Alice's public key>` to the base transaction properties.
3. It's `recipientAddress` instead of `recipient`
4. I think it would be good to include the most important parameters of the CCM in this graphic: `tokenID`, `amount`, `senderAddress` and `recipientAddress`. But it's up to you.

//TODO: Add link to certificate explanations
The Lisk interoperability solution allows blockchain to blockchain communication by using the paradigm of *cross-chain certification*.

> Basically, cross-chain certification means that information from one chain is submitted to another chain utilizing a signed object called a *certificate*.

Reviewer: ```suggestion
Basically, cross-chain certification means that information from one chain is authenticated and included in another chain utilizing a signed object called a *certificate*.
```

> The information is transmitted across chains by issuing a Cross-Chain Update transaction, aka CCU, including a certificate, that verifies the correctness of the data in the CCU.

Reviewer: ```suggestion
This information is transmitted across chains by issuing a Cross-Chain Update transaction, aka CCU.
```

> A CCU can contain multiple cross-chain messages.

Reviewer: ```suggestion
The CCU includes the certificate and the cross-chain messages coming from the sending chain.
```

> Cross-chain messages contain commands which can be executed on the opposite chain.

Reviewer: ```suggestion
Cross-chain messages contain cross-chain commands which are executed on the receiving chain.
```

[[ccc]]
Cross-chain command (CCC)::
Trigger of a state transition coming from another chain.
[[cert]]
Certificates::
Certificates contain information from finalized block headers that are signed by a large portion of validators from a certain chain, and thus authenticate a finalized state of that chain.
They are the fundamental components of the cross-chain certification paradigm for the Lisk ecosystem.

== Cross-chain network topology

The cross-chain network topology is organized as a **star topology** and describes arrangements of blockchains within the Lisk ecosystem.{fn_wiki_topology}

//TODO: Add link to network page, explaining network topology for Lisk blockchains

.Network topology of the Lisk blockchain ecosystem
image::understand-blockchain/interop/cross-chain topology.png[]

As shown by the image above, a Lisk blockchain can become interoperable to any other blockchain in the Lisk ecosystem including the Lisk Mainchain, simply by registering itself to the Lisk Mainchain.
All communication to other sidechains is routed through the Lisk Mainchain.

.What is the maximum of connected sidechains to the Lisk Mainchain?
[TIP]
====

> **The maximum of sidechains being simultaneously connected to the Lisk mainchain is currently around 3,5 million.**

Reviewer: I would remove this sentence, it is true in a sense, but maybe it is just confusing here. I would however change the following one as suggested

> **The maximum of sidechains being simultaneously connected to the Lisk mainchain is currently around 3,5 million.**
> 

Author: ```suggestion
```

> The number is only capped by the maximum number of CCU transactions, that the Lisk Mainchain can integrate per day.
> ====
> 
> == Mainchain & sidechains

Reviewer: ```suggestion
The maximum number of sidechains that can be registered in the Lisk ecosystem is only capped by the maximum number of CCU transactions that the Lisk Mainchain can integrate. In this sense, it is practically unlimited.
====

== Mainchain & sidechains
```

> As explained in section <<cross-chain-network-topology>> above, the Lisk blockchain network is structured in two general kinds of network particpants:

Reviewer: ```suggestion
As explained in section <<cross-chain-network-topology>> above, the Lisk blockchain network is structured in two general kinds of network participants:
```

> the *Lisk Mainchain*, and *sidechains*, which are connected to each other via the Mainchain.

Reviewer: I see what you mean, but I find the formulation a bit strange, to say that mainchain and sidechains are connected via the mainchain. Maybe we could just rewrite the first two sentences as smtg like:

In the Lisk ecosystem, sidechains are connected to each other via the mainchain. 

And then explain that all sidechains have to first register on the mainchain, then register the mainchain on the sidechain and then post a first CCU to start communicate with each other. I would even go as far as to have a bullet list point that enumerates all the steps as clearly as possible.

The *mainchain in that sense is a relayer for all cross-chain communication* happening in the network.

> The token used on the mainchain, LSK, has a special role in the ecosystem, too:

Reviewer: And crucially LSK is used to pay transaction fees on the mainchain, where CCUs from the sidechains have to be posted

All transaction fees in sidechains are paid in LSK by default, but it is possible to configure a sidechain token for the transaction fees.

> Apart from the special roles of the mainchain mentioned above, *the mainchain is not different to any other sidechain in the network.*

Reviewer: I find this sentence misleading, like, of course apart from this major difference there is no other difference. I would remove it

Author: Ofc there is difference between different blockchains, but not structurally. Any sidechain could become mainchain if the network decides on it, as far as I understand.

I will try to rephrase to make it clearer.

> The mainchain and each sidechain differ mainly from each other by the modules which are registered in the blockchain application.

Reviewer: Here I think the real difference is that a user cannot modify the mainchain protocol. In order to develop and deploy a dapp they have to spin up a sidechain (we are not a smart contract platform)

Author: This paragraph is comparing mainchain & sidechains structurally. The characteristics of blockchain applications compared to smart contract platforms will be covered on the `understand-blockchain/sdk/index.html` page

The mainchain only integrates a set of default modules, while sidechains normally have one or more additional modules registered, which provide functionality needed for their particular use case.

.The Role of the LSK Token
[NOTE]
====

> The LSK token is the only token that can be transferred to every chain within the Lisk ecosystem.

Reviewer: I don't think this is necessarily true. @gkoumout ?

It is the default token for transaction fees on sidechains, but it is possible to configure a sidechain token for the transaction fees.
As the LSK token is listed on several exchanges and there are fiat on-ramps available, it will in most cases be the initial token that a user acquires within the Lisk ecosystem.
Once a user possesses some LSK tokens, they can exchange them for other sidechain tokens, e.g., on a decentralized exchange (DEX) sidechain.
====

Because of the network topology, there are two different ways for cross-chain communication:

Sidechain-to-Mainchain (& vice versa)::
Involves creation of one CCU on the sending chain, that is posted to the receiving chain.
Sidechain-to-Sidechain::
Involves creation of two CCUs:
+
. CCU from sending chain to Mainchain
. CCU from Mainchain chain to receiving chain

.Sidechain-to-Sidechain communication via the Mainchain
image:understand-blockchain/interop/s2s.png[]

* On `sidechain 1`, three transactions are included, where each one emits one cross-chain message, denoted by `m1`, `m2`, and `m3`.

> * The color of a transaction or cross-chain message is always the one of the receiving chain, except for a cross-chain update transaction whose color is the one of the sending chain.

Reviewer: this should maybe go in the caption of the image

* All three cross-chain messages are delivered in one cross-chain update transaction, `CCU1`, to the mainchain, where `m1` and m3 are processed, but not `m2`.
* Later on, `m2` is delivered to sidechain 2 by a cross-chain update transaction, `CCU2`, from the mainchain to `sidechain 2`.
* This cross-chain update transaction contains an additional cross-chain message, `m4`, emitted by the transaction `t4` included in the mainchain.

== Example of Interoperability Usage
Let’s look at an example to get a better impression of the capabilities of our interoperability solution.
All the following steps described here can also be seen in the image below.

> Assume we have an exchange chain, a prediction market chain, and an oracle chain connected to the mainchain.

Reviewer: Maybe we can come up with a better example using one of the grant applications? Also, Penrose is not the most recent Nobel prize winner in physics anymore :)

Author: We can take that as an improvement suggestion for the future, when we have more time ti improve. For first version, the example is good enough imho.

Author: created a corresponding issue for this: https://github.com/LiskHQ/lisk-docs/issues/1625

. Subsequently, the betting tokens are then sent from the exchange chain to the prediction market chain via a cross-chain token transfer message.
. On the prediction market chain, the user bets on the winner of the Nobel Prize in Physics.
. After the announcement of the Nobel prize winner, the oracle chain sends the result to the prediction market chain via a custom cross-chain message.
. The user then receives their winnings as they made the correct guess.

.Lisk blockchain interoperability: Use case example
image::understand-blockchain/interop/high-level-overview-interoperability-fig_42x.png[]

Example of interoperability between the Lisk mainchain and three sidechains.

* The steps `2)`, `4)`, and `6)` are transactions performed within a single chain.
* The steps `1)`, `3)`, and `5)` are cross-chain messages.
* The cross-chain messages `3)` and `5)` are sidechain-to-sidechain cross-chain messages which are routed via the mainchain.
* The cross-chain message `1)` is a mainchain-to-sidechain cross-chain token transfer message.

Another example use case for Lisk interoperability was presented at Lisk.js 2021 by Alessandro Ricottone:

video::BTtLbhSgubA?t=827[youtube,500,300]

> .Additional educxational resources to learn about the Lisk interoperability solution

Reviewer: ```suggestion
.Additional educational resources to learn about the Lisk interoperability solution
```

They are the fundamental components of the cross-chain certification paradigm for the Lisk ecosystem.

== Cross-chain network topology

The cross-chain network topology is organized as a **star topology** and describes arrangements of blockchains within the Lisk ecosystem.{fn_wiki_topology}

//TODO: Add link to network page, explaining network topology for Lisk blockchains

.Network topology of the Lisk blockchain ecosystem
image::understand-blockchain/interop/cross-chain-topology.png[,500]

As shown by the image above, a Lisk blockchain can become interoperable to any other blockchain in the Lisk ecosystem including the Lisk Mainchain, simply by registering itself to the mainchain.
All communication to other sidechains is routed through the mainchain.

.What is the maximum of connected sidechains to the Lisk Mainchain?
[TIP]
====
The maximum number of sidechains that can be registered in the Lisk ecosystem is only capped by the maximum number of CCU transactions that the Lisk Mainchain can integrate.
In this sense, it is practically unlimited.
====

== Mainchain & sidechains

> As explained in section <<cross-chain-network-topology>> above, the Lisk blockchain network is structured in two general kinds of network participants:

Author: ```suggestion
As explained in section <<cross-chain-network-topology>> above, the Lisk blockchain network is structured in two types of network participants:
```

the *Lisk Mainchain*, and *sidechains*.

In the Lisk ecosystem, sidechains are connected to each other via the mainchain.

To become a sidechain in the Lisk ecosystem, a blockchain application needs to perform the following three simple steps:

. Register on the mainchain.
. Register the mainchain on the sidechain.
. Post a first CCU to start communicate with each other.

NOTE: Because sidechains communicate through the mainchain with other sidechains, they only need to maintain one connection to the mainchain, which will significantly reduce both the complexity and the point-of-failure.

****

> Each step of the lifecycle of a sidechain is explained in further detail on the page xref:{url_understand_sidechainlifecycle}[].

Author: the content will be added in another task, see PR https://github.com/LiskHQ/lisk-docs/pull/1632

****

The mainchain forwards all cross-chain communication happening between registered sidechains in the network.

> Structurally, the mainchain is not any different from any other sidechain.

Author: Oh, I thought it is actually already possible to do so, and communicating over the Lisk mainchain is rather optinal than mandatory. But if this is not the case, then I agree we should probably remove or update this section.

> Structurally, the mainchain is not any different from any other sidechain.

Reviewer 2: No, it will definitely not be possible with the Sapphire phase SDK.

. The LSK tokens are then swapped for the betting tokens.
. Subsequently, the betting tokens are then sent from the exchange chain to the prediction market chain via a cross-chain token transfer message.
. On the prediction market chain, the user bets on the winner of the Nobel Prize in Physics.
. After the announcement of the Nobel prize winner, the oracle chain sends the result to the prediction market chain via a custom cross-chain message.
. The user then receives their winnings as they made the correct guess.

.Lisk blockchain interoperability: Use case example
image::understand-blockchain/interop/high-level-overview-interoperability-fig_42x.png[,600]

Example of interoperability between the Lisk mainchain and three sidechains.

* The steps `2)`, `4)`, and `6)` are transactions performed within a single chain.
* The steps `1)`, `3)`, and `5)` are cross-chain messages.
* The cross-chain messages `3)` and `5)` are sidechain-to-sidechain cross-chain messages which are routed via the mainchain.
* The cross-chain message `1)` is a mainchain-to-sidechain cross-chain token transfer message.

> Another example use case for Lisk interoperability was presented at Lisk.js 2021 by Alessandro Ricottone:

Author: Good point, I commented the section out.

[NOTE]
====
The maximum number of sidechains that can be registered in the Lisk ecosystem is only capped by the maximum number of CCU transactions that the Lisk Mainchain can process.
In this sense, it is practically unlimited.
====

== Mainchain & sidechains

As explained in section <<cross-chain-network-topology>> above, the Lisk blockchain network is structured in two types of network participants:

* the *Lisk Mainchain*:  Follows the Lisk Mainchain protocol and cannot be modified by a blockchain application developer, and
* *sidechains*: Blockchain applications built with the Lisk SDK.
To become part of the Lisk blockchain ecosystem, they need to follow the Lisk protocol, for example, they need to include the Lisk Interoperability module.
All additional parts of the sidechain protocol are defined by the blockchain application developer.

In the Lisk ecosystem, sidechains are connected to each other via the mainchain.
Because sidechains communicate through the mainchain with other sidechains, they only need to maintain one connection to the mainchain, which will significantly reduce both the complexity and the point-of-failure.

> To become a sidechain in the Lisk ecosystem, a blockchain application needs to perform the following three simple steps:

Reviewer 3: It looks like the third step below this statement is missing here..?
"To become a sidechain in the Lisk ecosystem, a blockchain application needs to perform the following three simple steps":

Author: ```suggestion
To become a sidechain in the Lisk ecosystem, a blockchain application needs to perform the following two simple steps:
```

ah yes. I removed the third step, as discussed with @AndreasKendziorra

image::understand-blockchain/interop/s2s.png[,600]

* On `sidechain X`, three transactions are included, where each one emits one cross-chain message, denoted by `CCM1`, `CCM2`, and `CCM3`.
* All three cross-chain messages are delivered in one cross-chain update transaction, `CCU1`, to the mainchain, where `CCM1` and `CCM3` are processed, but not `CCM2`.
* Later on, `CCM2` is delivered to `sidechain Y` by a cross-chain update transaction, `CCU2`, from the mainchain to `sidechain Y`.
* This cross-chain update transaction contains an additional cross-chain message, `CCM4`, emitted by the transaction `CCT4` included in the mainchain.

****
The cross-chain communication is explained in further detail on the page xref:{url_understand_ccc}[].
****

== Example of interoperability usage
Let’s look at an example to get a better impression of the capabilities of our interoperability solution.
All the following steps described here can also be seen in the image below.

Assume we have an exchange chain, a prediction market chain, and an oracle chain connected to the mainchain.

> Then, a user story could look like this:

Reviewer 3: ```suggestion
Then, a user story could look like the example described below:
```

== Mainchain & sidechains

As explained in section <<cross-chain-network-topology>> above, the Lisk blockchain network is structured in two types of network participants:

* the *Lisk Mainchain*:  Follows the Lisk Mainchain protocol and cannot be modified by a blockchain application developer, and
* *sidechains*: Blockchain applications built with the Lisk SDK.
To become part of the Lisk blockchain ecosystem, they need to follow the Lisk protocol, for example, they need to include the Lisk Interoperability module.
All additional parts of the sidechain protocol are defined by the blockchain application developer.

In the Lisk ecosystem, sidechains are connected to each other via the mainchain.
Because sidechains communicate through the mainchain with other sidechains, they only need to maintain one connection to the mainchain, which will significantly reduce both the complexity and the point-of-failure.

To become a sidechain in the Lisk ecosystem, a blockchain application needs to perform the following three simple steps:

. Register on the mainchain.
. Register the mainchain on the sidechain.

> . Post a first CCU to start communicate with each other.

Reviewer 2: To become a "sidechain with status active", a CCU with at least on CCM needs to be send. But for simplicity one could also remove this step here completely. Then appears even less complex.
